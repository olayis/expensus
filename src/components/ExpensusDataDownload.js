import React from 'react';
import moment from 'moment';
import { CSVLink } from 'react-csv';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faFileCsv } from '@fortawesome/free-solid-svg-icons';
import { faArrowAltCircleDown } from '@fortawesome/free-regular-svg-icons';
import logEvent from '../actions/logEvent';

const ExpensusDataDownload = (props) => {
  return (
    <div className='component'>
      <h2 className='heading-secondary'>
        <FontAwesomeIcon icon={faFileCsv} className='fa-icon' /> Download Data
      </h2>
      <div>
        <p className='component__message'>
          Click on the download icon to download your expenses in CSV format.
          Expenses downloaded are based on filters.
        </p>
        <div className='component__content'>
          <CSVLink
            data={props.expenses.map((expense, index) => ({
              ...expense,
              id: index + 1,
              amount: expense.amount / 100,
              createdAt: moment(expense.createdAt).format('DD/MM/YYYY'),
            }))}
            aria-label='CSV Download Link'
            filename={'expenses.csv'}
            onClick={() =>
              logEvent('download_expense', { info: 'User downloaded data' })
            }
          >
            <FontAwesomeIcon
              icon={faArrowAltCircleDown}
              className='component__illustration'
            />
          </CSVLink>
        </div>
      </div>
    </div>
  );
};

export default ExpensusDataDownload;
