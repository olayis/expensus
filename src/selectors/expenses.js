// Get Visible expenses
import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      switch (sortBy) {
        case 'date(recent)':
          return a.createdAt < b.createdAt ? 1 : -1;

        case 'date(oldest)':
          return a.createdAt > b.createdAt ? 1 : -1;

        case 'amount(highest)':
          return a.amount < b.amount ? 1 : -1;

        case 'amount(lowest)':
          return a.amount > b.amount ? 1 : -1;

        case 'description(a_to_z)':
          return a.description.localeCompare(b.description);

        case 'description(z_to_a)':
          return b.description.localeCompare(a.description);

        case 'category':
          return a.category.localeCompare(b.category);
      }
    });
};
