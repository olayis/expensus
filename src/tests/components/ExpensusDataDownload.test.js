import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpensusDataDownload from '../../components/ExpensusDataDownload';

test('should render ExpensusDataDownload', () => {
  const wrapper = shallow(<ExpensusDataDownload expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});
