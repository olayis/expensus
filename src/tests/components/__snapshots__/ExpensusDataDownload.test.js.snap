// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensusDataDownload 1`] = `
<div
  className="component"
>
  <h2
    className="heading-secondary"
  >
    <FontAwesomeIcon
      border={false}
      className="fa-icon"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            384,
            512,
            Array [],
            "f6dd",
            "M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm-96 144c0 4.42-3.58 8-8 8h-8c-8.84 0-16 7.16-16 16v32c0 8.84 7.16 16 16 16h8c4.42 0 8 3.58 8 8v16c0 4.42-3.58 8-8 8h-8c-26.51 0-48-21.49-48-48v-32c0-26.51 21.49-48 48-48h8c4.42 0 8 3.58 8 8v16zm44.27 104H160c-4.42 0-8-3.58-8-8v-16c0-4.42 3.58-8 8-8h12.27c5.95 0 10.41-3.5 10.41-6.62 0-1.3-.75-2.66-2.12-3.84l-21.89-18.77c-8.47-7.22-13.33-17.48-13.33-28.14 0-21.3 19.02-38.62 42.41-38.62H200c4.42 0 8 3.58 8 8v16c0 4.42-3.58 8-8 8h-12.27c-5.95 0-10.41 3.5-10.41 6.62 0 1.3.75 2.66 2.12 3.84l21.89 18.77c8.47 7.22 13.33 17.48 13.33 28.14.01 21.29-19 38.62-42.39 38.62zM256 264v20.8c0 20.27 5.7 40.17 16 56.88 10.3-16.7 16-36.61 16-56.88V264c0-4.42 3.58-8 8-8h16c4.42 0 8 3.58 8 8v20.8c0 35.48-12.88 68.89-36.28 94.09-3.02 3.25-7.27 5.11-11.72 5.11s-8.7-1.86-11.72-5.11c-23.4-25.2-36.28-58.61-36.28-94.09V264c0-4.42 3.58-8 8-8h16c4.42 0 8 3.58 8 8zm121-159L279.1 7c-4.5-4.5-10.6-7-17-7H256v128h128v-6.1c0-6.3-2.5-12.4-7-16.9z",
          ],
          "iconName": "file-csv",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
     Download Data
  </h2>
  <div>
    <p
      className="component__message"
    >
      Click on the download icon to download your expenses in CSV format. Expenses downloaded are based on filters.
    </p>
    <div
      className="component__content"
    >
      <CSVLink
        aria-label="CSV Download Link"
        asyncOnClick={false}
        data={
          Array [
            Object {
              "amount": 20,
              "category": "Bills & Utilities",
              "createdAt": "01/01/1970",
              "description": "Data",
              "id": 1,
              "note": "",
            },
            Object {
              "amount": 350,
              "category": "Rent",
              "createdAt": "27/12/1969",
              "description": "Home Rentage",
              "id": 2,
              "note": "",
            },
            Object {
              "amount": 80,
              "category": "Maintenance",
              "createdAt": "13/01/1970",
              "description": "Room Repainting",
              "id": 3,
              "note": "",
            },
            Object {
              "amount": 10,
              "category": "Maintenance",
              "createdAt": "11/01/1970",
              "description": "Water Pipes Repair",
              "id": 4,
              "note": "",
            },
          ]
        }
        filename="expenses.csv"
        separator=","
        target="_blank"
        uFEFF={true}
      >
        <FontAwesomeIcon
          border={false}
          className="component__illustration"
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f358",
                "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200zm-32-316v116h-67c-10.7 0-16 12.9-8.5 20.5l99 99c4.7 4.7 12.3 4.7 17 0l99-99c7.6-7.6 2.2-20.5-8.5-20.5h-67V140c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12z",
              ],
              "iconName": "arrow-alt-circle-down",
              "prefix": "far",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </CSVLink>
    </div>
  </div>
</div>
`;
